### PR TITLE
nano/tiny deepseek moe training

### PR DESCRIPTION
**Pull request by philei-tt**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/41643 

### Problem description
We need MoE model we're going to optimize out for our MoE training effort. 

#### Summary
- Implement a nano/tiny-scale DeepSeek-V3 architecture in ttml Python, featuring Multi-head Latent Attention (MLA) and Mixture-of-Experts (MoE) with DeepSeek-style sigmoid routing
Add FLOPs-per-token calculation and MFU reporting for DeepSeek models in the training loop
- Provide two model configs: nano (~35M params, for fast iteration) and tiny (~1.75B total / 0.47B active, for more realistic MoE training)
- Note: this is really basic naive implementation lacking efficient routing, sparse moe and fused sdpa.

#### Architecture highlights
**Multi-head Latent Attention (MLA) — compresses KV cache via low-rank joint projection:**
- Q path: wq_a (down-project) -> RMSNorm -> wq_b (up-project) -> split nope/rope -> RoPE on rope subspace only -> concat
- KV path: wkv_a -> split latent/k_pe -> RMSNorm -> wkv_b (up-project) -> split k_nope/v + RoPE(k_pe) broadcast -> concat
- Uses composite SDPA to support v_head_dim != qk_head_dim (architecturally faithful to the original DeepSeek design)

**Mixture-of-Experts (MoE) -- DeepSeek-V3-style routing with all computation on device:**
- Sigmoid scoring -> top-k expert selection via ttnn.topk (no CPU sync during forward)
- Per-expert binary masks built on device with ttnn.eq / ttnn.sum / ttnn.gt
- Score normalization across selected experts (matching reference DeepSeek Gate)
- Dense masking execution: each expert processes full input, masked by selection
- Shared expert always active
- Auxiliary-loss-free load balancing via expert bias buffer

#### Modifications / New files
- New model location: `tt-train/sources/ttml/ttml/models/deepseek/ `
- New model configs: `configs/model_configs/moe/nano_deepseek.yaml`, `configs/model_configs/moe/tiny_deepseek.yaml`
- New training configs:  `configs/training_configs/training_shakespeare_nano_deepseek.yaml` (B=32), `configs/training_configs/training_shakespeare_tiny_deepseek.yaml` (B=8)
- Update `train_nanogpt.py` -- Added model_type="deepseek" support, TPS/TFLOPS/MFU reporting, device peak TFLOPS detection
- Update `scaled_dot_product_attention.cpp` -- Relaxed validate_qkv_shapes to allow v_dim != q_dim for composite SDPA path (fused kernel retains its own stricter checks in device ops)

#### Performance / Accuracy
`configs/training_configs/training_shakespeare_nano_deepseek.yaml`

~270ms / step, 2.6% mfu
```
   - Model: 8 layers, 512 embd, 8 heads
   - Total parameters: 32,509,440
   - FLOPs per token: 125,779,968 (0.13G)
   - Overall DRAM Peak: 5398.28 MB, Final DRAM Usage: 534.88 MB

Step: 3, Loss: 4.062500, Time: 275.08 ms, TPS: 29780, TFLOPS: 3.75, MFU: 2.5%
Step: 4, Loss: 3.765625, Time: 275.67 ms, TPS: 29717, TFLOPS: 3.74, MFU: 2.5%
Step: 5, Loss: 3.656250, Time: 275.56 ms, TPS: 29728, TFLOPS: 3.74, MFU: 2.5%
Step: 6, Loss: 3.515625, Time: 275.65 ms, TPS: 29718, TFLOPS: 3.74, MFU: 2.5%
```

**Comparison to GPU baseline**
[GPU baseline](https://github.com/philei-tt/tt-train_nanoGPT-gpu-baseline/blob/master/train_deepseek_torch.py) implements the same shakespeare, next-token-prediction pretraining with identical dtypes (bf16 for model, fp32 for adamw) and identical optimizer config, and it's based on the original implementation [from here](https://github.com/deepseek-ai/DeepSeek-V3/blob/main/inference/model.py)

Comparison vs RTX5080

image

`configs/training_configs/training_shakespeare_tiny_deepseek.yaml`
```
   - Model: 16 layers, 1536 embd, 12 heads
   - Total parameters: 1,754,789,376
   - FLOPs per token: 2,355,757,056 (2.36G)
   - Overall DRAM Peak: 29617.71 MB, Final DRAM Usage: 26850.12 MB

Step: 3, Loss: 5.406250, Time: 3395.64 ms, TPS: 1206, TFLOPS: 2.84, MFU: 1.9%
Step: 4, Loss: 4.593750, Time: 3394.78 ms, TPS: 1207, TFLOPS: 2.84, MFU: 1.9%
Step: 5, Loss: 4.468750, Time: 3395.26 ms, TPS: 1206, TFLOPS: 2.84, MFU: 1.9%
Step: 6, Loss: 4.062500, Time: 3395.79 ms, TPS: 1206, TFLOPS: 2.84, MFU: 1.9%
```

UPD: 
Added "real" deepseek moe 16B model config. 

UPD2: 
Fixed bug @nathan-TT found with ttnn.eq routing only to expert 0. Also fixed bug with weights normalization not propagating correct gradient, and added tests for that. New loss: 

image

This also improved mfu and tps for nano deepseek: 
```
Step: 5000, Loss: 0.213867, Time: 240.28 ms, TPS: 34094, TFLOPS: 4.29, MFU: 2.9%
```

and tiny deepseek: 
```
Step: 4, Loss: 4.031250, Time: 2358.78 ms, TPS: 1736, TFLOPS: 4.09, MFU: 2.8%
```

### Checklist

- [ ] [Sanity tests](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:philei/train-nano-deepseek-moe)

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:philei/train-nano-deepseek-moe)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:philei/train-nano-deepseek-moe)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:philei/train-nano-deepseek-moe)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:philei/train-nano-deepseek-moe)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:philei/train-nano-deepseek-moe)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:philei/train-nano-deepseek-moe)

### CODE REVIEW

File: tt-train/sources/ttml/ttml/models/deepseek/__init__.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+#
+# SPDX-License-Identifier: Apache-2.0
+
+"""Nano DeepSeek-V3 model package.
+
+Implements the DeepSeek-V3 architecture (671B total / 37B active) at nano scale:
+  - Multi-head Latent Attention (MLA) with low-rank Q/KV compression
+  - DeepSeek-style Mixture of Experts (MoE) with group routing
+  - SwiGLU activation in all feed-forward layers
+  - RoPE applied only to a subspace of Q/K heads
```
[the remark this answers is not in the document]
Comment (philei-tt) [reply tt-train/sources/ttml/ttml/models/deepseek/__init__.py]: Good catch

File: tt-train/sources/ttml/ttml/models/deepseek/moe.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+        Call between training steps (outside forward). Only CPU interaction.
+        """
+        bias_val = self._expert_bias.tensor.get_value()
+        fp32 = ttnn.typecast(bias_val, ttnn.DataType.FLOAT32)
+        bias_np = np.array(ttnn.to_torch(ttnn.from_device(fp32)).numpy(), dtype=np.float32).reshape(1, 1, 1, -1)
+
+        delta = np.random.normal(0, coeff, size=bias_np.shape).astype(np.float32)
+        delta -= delta.mean()
+        bias_np = bias_np + delta
```
[the remark this answers is not in the document]
Comment (philei-tt) [reply tt-train/sources/ttml/ttml/models/deepseek/moe.py]: Good catch

File: tt-train/sources/ttml/ttml/models/deepseek/moe.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+        # (reference DeepSeek only normalizes for sigmoid).
+
+        expert_masks = {}  # expert_idx -> ttnn mask [B, 1, S, 1]
+        denom = None  # [B, 1, S, 1] running sum of selected scores
+
+        for expert_idx in range(self.num_experts):
+            match = ttnn.eq(topk_indices, float(expert_idx))
+            match_f = ttnn.typecast(match, ttnn.DataType.BFLOAT16)
+            match_any = ttnn.sum(match_f, dim=-1, keepdim=True)  # [B, 1, S, 1]
+            mask_narrow = ttnn.gt(match_any, 0.0)  # [B, 1, S, 1]
```
[the remark this answers is not in the document]
Comment (philei-tt) [reply tt-train/sources/ttml/ttml/models/deepseek/moe.py]: ttnn.eq doesn't work with integer

File: tt-train/sources/ttml/ttml/models/deepseek/__init__.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        x = self.tok_emb(tokens)
+
+        for block in self.blocks:
+            if self.config.runner_type == RunnerType.MemoryEfficient:
+                x = memory_efficient_runner(block, x, mask)
+            else:
+                x = block(x, mask)
+
+        x = self.norm(x)
+        return self.head(x)
```
[the remark this answers is not in the document]
Comment (philei-tt) [reply tt-train/sources/ttml/ttml/models/deepseek/__init__.py:109]: Maybe

File: tt-train/sources/examples/nano_gpt/train_nanogpt.py [hunk 5]
```diff
@@ -1309,12 +1403,22 @@ def main():
         gradient_accumulator = GradientAccumulator(training_config.gradient_accumulation_steps)
         global_step = start_step
 
+        # Compute peak device TFLOPS for MFU calculation
+        peak_tflops = 0.0
+        if flops_per_token > 0:
+            try:
+                peak_tflops = get_device_peak_tflops_bf16()
+                print(f"  - Device peak: {peak_tflops:.1f} TFLOPS (bf16)")
+            except Exception:
```
[the remark this answers is not in the document]
Comment (philei-tt) [reply tt-train/sources/examples/nano_gpt/train_nanogpt.py]: Fixed

File: tt-train/sources/examples/nano_gpt/train_nanogpt.py [hunk 6]
```diff
@@ -45,18 +45,39 @@
     LlamaConfig,
     LlamaRopeScalingConfig,
 )
+from ttml.models.deepseek import (
+    DeepSeek,
+    DeepSeekConfig,
+    calculate_flops_per_token as deepseek_flops_per_token,
+)
```
[the remark this answers is not in the document]
Comment (philei-tt) [reply tt-train/sources/examples/nano_gpt/train_nanogpt.py:51]: Fixed

File: tt-train/sources/ttml/ttml/models/deepseek/__init__.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    ) -> ttml.autograd.Tensor:
+        """Forward pass.
+
+        Args:
+            tokens: Token IDs [B, 1, 1, S]
+            mask: Causal attention mask [1, 1, S, S]
+
+        Returns:
+            Logits [B, 1, S, vocab_size]
+        """
```
[the remark this answers is not in the document]
Comment (philei-tt) [reply tt-train/sources/ttml/ttml/models/deepseek/__init__.py:95]: Fixed

File: tt-train/sources/ttml/ttml/models/deepseek/autograd_ops.py [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+class Slice(ttml.autograd.Function):
+    """Autograd-aware slice (ttnn.slice has no backward).
+
+    Forward: extract a sub-tensor from input.
+    Backward: place gradient into zeros at the sliced position.
+    """
+
+    @staticmethod
+    def forward(ctx, input, start, end):
+        ctx.save_for_backward(input)
```
Comment (zbaczewskiTT) [inline tt-train/sources/ttml/ttml/models/deepseek/autograd_ops.py]: We don't have to save the whole input tensor for backward; only shape, start and end.

File: tt-train/sources/ttml/ttml/models/deepseek/autograd_ops.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+                z_before = ttnn.zeros(z_shape, dtype=result.dtype, layout=result.layout, device=device)
+                parts.append(z_before)
+
+            parts.append(result)
+
+            if after > 0:
+                z_shape = list(cur_shape)
+                z_shape[dim] = after
+                z_after = ttnn.zeros(z_shape, dtype=result.dtype, layout=result.layout, device=device)
+                parts.append(z_after)
```
Comment (zbaczewskiTT) [inline tt-train/sources/ttml/ttml/models/deepseek/autograd_ops.py:60]: Ideally we'd allocate a single tensor of input_shape shape and only write to the [start:end] slice; but it may be out of the scope for this PR

File: tt-train/sources/ttml/ttml/models/deepseek/autograd_ops.py [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+    Backward: split gradient and distribute to each input.
+    """
+
+    @staticmethod
+    def forward(ctx, dim, *tensors):
+        sizes = [list(t.get_value().shape)[dim] for t in tensors]
+        ctx.sizes = sizes
+        ctx.dim = dim
+        ctx.num_tensors = len(tensors)
+        ctx.save_for_backward(*tensors)
```
Comment (zbaczewskiTT) [inline tt-train/sources/ttml/ttml/models/deepseek/autograd_ops.py]: Why do we need to save it for backward?

File: tt-train/sources/ttml/ttml/models/deepseek/autograd_ops.py [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+    """Autograd-aware sigmoid.
+
+    Forward: y = sigmoid(x)
+    Backward: dL/dx = dL/dy * y * (1 - y)
+    """
+
+    @staticmethod
+    def forward(ctx, input):
+        y = ttnn.sigmoid(input.get_value())
+        ctx.save_for_backward(input)
```
Comment (zbaczewskiTT) [inline tt-train/sources/ttml/ttml/models/deepseek/autograd_ops.py]: Again, why are we saving it for backward?

File: tt-train/sources/ttml/ttml/models/deepseek/autograd_ops.py [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+    """Autograd-aware softmax on last dim.
+
+    Forward: y = softmax(x, dim=-1)
+    Backward: uses ttnn.moreh_softmax_backward (fused kernel)
+    """
+
+    @staticmethod
+    def forward(ctx, input):
+        y = ttnn.softmax(input.get_value(), dim=-1)
+        ctx.save_for_backward(input)
```
Comment (zbaczewskiTT) [inline tt-train/sources/ttml/ttml/models/deepseek/autograd_ops.py]: This one too.

File: tt-train/sources/ttml/ttml/models/deepseek/autograd_ops.py [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+      Forward:  reshape to (B, S, H, D) then transpose dims 1,2
+      Backward: transpose dims 1,2 then reshape back
+    """
+
+    @staticmethod
+    def forward(ctx, input, num_heads):
+        val = input.get_value()
+        B, _, S, HD = list(val.shape)
+        head_dim = HD // num_heads
+        ctx.save_for_backward(input)
```
Comment (zbaczewskiTT) [inline tt-train/sources/ttml/ttml/models/deepseek/autograd_ops.py]: Same problem.

File: tt-train/sources/ttml/ttml/models/deepseek/autograd_ops.py (hunk 8 above)
Comment (philei-tt) [reply tt-train/sources/ttml/ttml/models/deepseek/autograd_ops.py]: Good catch

File: tt-train/sources/ttml/ttml/models/deepseek/moe.py [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+        self.experts = ModuleList([Expert(config.dim, config.moe_inter_dim) for _ in range(config.n_routed_experts)])
+
+        # Shared expert(s)
+        if config.n_shared_experts > 0:
+            self.shared_experts = DeepSeekMLP(config.dim, config.n_shared_experts * config.moe_inter_dim)
+        else:
+            self.shared_experts = None
+
+        # Load balancing bias buffer (on device, TILE layout)
+        bias_np = np.zeros((1, 1, 1, config.n_routed_experts), dtype=np.float32)
```
Comment (dyurshevichTT) [inline tt-train/sources/ttml/ttml/models/deepseek/moe.py]: nit: ttnn.zeros

File: tt-train/sources/ttml/ttml/models/deepseek/moe.py [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+        self._expert_bias = Buffer(ttml.autograd.Tensor.from_numpy(bias_np, ttnn.Layout.TILE))
+
+        # On-device accumulator for token counts per expert [1, 1, 1, num_experts]
+        # Each forward adds to this; update_expert_bias reads and resets it.
+        counts_np = np.zeros((1, 1, 1, config.n_routed_experts), dtype=np.float32)
```
Comment (dyurshevichTT) [inline tt-train/sources/ttml/ttml/models/deepseek/moe.py]: and here

**Review by dyurshevichTT — APPROVED** [review]
LGTM!

File: tt-train/sources/ttml/ttml/models/deepseek/moe.py [hunk 16]
```diff
@@ -0,0 +1,263 @@
...
+        super().__init__()
+
+        # ttnn.topk returns indices as bf16 which represents integers exactly up to 256.
+        # Beyond that, ttnn.eq(topk_indices, float(expert_idx)) may match wrong experts.
+        if config.n_routed_experts > 256:
+            raise ValueError(
+                f"n_routed_experts={config.n_routed_experts} exceeds 256, the maximum "
+                f"integer exactly representable in bf16. On-device routing via "
+                f"ttnn.topk + ttnn.eq would produce incorrect expert masks."
+            )
```
[the remark this answers is not in the document]
Comment (philei-tt) [reply tt-train/sources/ttml/ttml/models/deepseek/moe.py]: Stupid copilot

**Review by vmelnykovTT — APPROVED** [review]
LGTM!

**Review by zbaczewskiTT — APPROVED** [review]

**Review by ayerofieiev-tt — APPROVED** [review]

**Review by dpopovTT — APPROVED** [review]

**Review by blozano-tt — APPROVED** [review]
